packages/library-sync/src/commonMain/kotlin/io/realm/kotlin/mongodb/internal/KtorWebSocketTransport.kt

```diff
@@ -0,0 +1,322 @@
+                                            if (shouldCloseSocket) {
+                                                closeWebsocket()
+                                            }
+                                        }
+
+                                        is Frame.Close -> {
+                                            // It's important to rely properly the error code from the server.
+                                            // The server will report auth errors (and a few other error types)
+                                            // as websocket application-level errors after establishing the socket, rather than failing at the HTTP layer.
+                                            // since the websocket spec does not allow the HTTP status code from the response to be
+                                            // passed back to the client from the websocket implementation (example instruct a refresh token
+                                            // via a 401 HTTP response is not possible) see https://jira.mongodb.org/browse/BAAS-10531.
+                                            // In order to provide a reasonable response that the Sync Client can react upon, the private range of websocket close status codes
+                                            // 4000-4999, can be used to return a more specific error.
+                                            val errorCode: WebsocketErrorCode =
+                                                frame.readReason()?.code?.toInt()
+                                                    ?.let { code -> WebsocketErrorCode.of(code) }
+                                                    ?: WebsocketErrorCode.RLM_ERR_WEBSOCKET_OK
+                                            val reason: String = frame.readReason()?.toString()
+                                                ?: "Received Close from Websocket server"
+
+                                            observer.onClose(true, errorCode, reason)
+                                        }
+
+                                        is Frame.Text -> {
+                                            logger.warn("Received unexpected text WebSocket message ${frame.readText()}")
+                                        }
+
+                                        // Raw WebSocket Frames (i.e Frame.Ping & Frame.Pong) are handled automatically with the client
+                                        // (Core doesn't care about these in the new API)
+                                        else -> {
+                                            logger.warn("Received unexpected message from server, Frame type = ${frame.frameType}")
+                                        }
+                                    }
+                                }
+                            }
+                            websocketJob.join() // otherwise the client will send end the session and send a Close
```

Comment:
The WebSocket opened in https://github.com/realm/realm-kotlin/pull/1528/files/7133797711950bfe1d7eec2cb30ad696823879bb#diff-2270f40a8e4fabaacf4f2791bbe75dee7428db4c09f84fc606ab90aaa99a78e1R119 will send a `close Frame` if we don't join the `websocketJob` this will maintain the scope/socket open until an error happens or core decides to close the socket